lib/Common/Memory/Recycler.cpp

```diff
@@ -5873,54 +5979,127 @@ Recycler::DoBackgroundWork(bool forceForeground)
+            ((!CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) ||!this->AllocationsDuringconcurrentSweepTurnedOn()) && this->collectionState == CollectionStateConcurrentSweep))
+#endif
         {
-            // Zero the queued pages first so they are available to be allocated
-            recyclerPageAllocator.BackgroundZeroQueuedPages();
-            recyclerLargeBlockPageAllocator.BackgroundZeroQueuedPages();
+            RECYCLER_PROFILE_EXEC_BACKGROUND_BEGIN(this, Js::ConcurrentSweepPhase);
+            GCETW_INTERNAL(GC_START, (this, ETWEvent_ConcurrentSweep));
+            GCETW(GC_BACKGROUNDZEROPAGE_START, (this));
+
+#if ENABLE_BACKGROUND_PAGE_ZEROING
+            if (CONFIG_FLAG(EnableBGFreeZero))
+            {
+                // Zero the queued pages first so they are available to be allocated
+                recyclerPageAllocator.BackgroundZeroQueuedPages();
+                recyclerLargeBlockPageAllocator.BackgroundZeroQueuedPages();
 #ifdef RECYCLER_WRITE_BARRIER_ALLOC_SEPARATE_PAGE
-            recyclerWithBarrierPageAllocator.BackgroundZeroQueuedPages();
+                recyclerWithBarrierPageAllocator.BackgroundZeroQueuedPages();
 #endif
-        }
+            }
 #endif
 
-        GCETW(GC_BACKGROUNDZEROPAGE_STOP, (this));
-        GCETW(GC_BACKGROUNDSWEEP_START, (this));
+            GCETW(GC_BACKGROUNDZEROPAGE_STOP, (this));
+            GCETW(GC_BACKGROUNDSWEEP_START, (this));
 
-        Assert(this->recyclerSweep != nullptr);
-        this->recyclerSweep->BackgroundSweep();
+            Assert(this->recyclerSweep != nullptr);
+            this->recyclerSweep->BackgroundSweep();
 
-        uint sweptBytes = 0;
+            // If allocations were allowed during concurrent sweep then the allocableHeapBlock lists still needs to be swept so we
+            // will remain in CollectionStateConcurrentSweepPass1Wait state.
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+            if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) && this->AllocationsDuringconcurrentSweepTurnedOn())
+            {
+                this->collectionState = CollectionStateConcurrentSweepPass1Wait;
+            }
+#endif
+        }
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+        if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc))
+        {
+            if (this->collectionState == CollectionStateConcurrentSweepPass2)
+            {
+#ifdef RECYCLER_TRACE
+                if (this->GetRecyclerFlagsTable().Trace.IsEnabled(Js::ConcurrentSweepPhase))
+                {
+                    Output::Print(_u("[GC #%d] Finishing Sweep Pass2 on background thread. \n"), this->collectionCount);
+                }
+#endif
+#if ENABLE_BACKGROUND_PAGE_ZEROING
+                if (CONFIG_FLAG(EnableBGFreeZero))
+                {
+                    // Drain the zero queue again as we might have free more during sweep
+                    // in the background
+                    GCETW(GC_BACKGROUNDZEROPAGE_START, (this));
```

Comment:
Why are you doing this before we finish the sweep? You're going to do this anyway below, no?

Reply:
Even for one pass sweep we try to do this before and after the sweep to make any queued pages available for allocation. I wanted to be consistent with that. Let me know if you think otherwise.

---
In reply to: [148171629](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148171629) [](ancestors = 148171629)